containers/unit_tests/TestDualView.hpp

```diff
@@ -211,3 +211,355 @@ TEST(TEST_CATEGORY, dualview_deep_copy) {
+    typedef typename ViewType::t_dev::memory_space::execution_space
+        t_dev_exec_space;
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<t_dev_exec_space>(0, n),
+        SumViewEntriesFunctor<scalar_type, typename ViewType::t_dev>(a.d_view),
+        a_d_sum);
+    ASSERT_EQ(a_d_sum, sum_total);
+
+    // Check host view is synced as expected
+    scalar_type a_h_sum = 0;
+    for (size_t i = 0; i < a.h_view.extent(0); ++i)
+      for (size_t j = 0; j < a.h_view.extent(1); ++j) {
+        a_h_sum += a.h_view(i, j);
+      }
+
+    ASSERT_EQ(a_h_sum, sum_total);
+
+    // Test deep_copy
+    Kokkos::deep_copy(b, a);
+    b.template sync<typename ViewType::host_mirror_space>();
+
+    // Perform same checks on b as done on a
+    // Check device view is initialized as expected
+    scalar_type b_d_sum = 0;
+    // Execute on the execution_space associated with t_dev's memory space
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<t_dev_exec_space>(0, n),
+        SumViewEntriesFunctor<scalar_type, typename ViewType::t_dev>(b.d_view),
+        b_d_sum);
+    ASSERT_EQ(b_d_sum, sum_total);
+
+    // Check host view is synced as expected
+    scalar_type b_h_sum = 0;
+    for (size_t i = 0; i < b.h_view.extent(0); ++i)
+      for (size_t j = 0; j < b.h_view.extent(1); ++j) {
+        b_h_sum += b.h_view(i, j);
+      }
+
+    ASSERT_EQ(b_h_sum, sum_total);
+
+  }  // end run_me
+
+  test_dual_view_deep_copy() {
+    run_me<Kokkos::DualView<Scalar**, Kokkos::LayoutLeft, Device> >();
+  }
+};
+
+template <typename Scalar, class Device>
+struct test_dualview_resize {
+  typedef Scalar scalar_type;
+  typedef Device execution_space;
+
+  template <typename ViewType>
+  void run_me() {
+
+    const unsigned int n      = 10;
+    const unsigned int m      = 5;
+    const unsigned int factor = 2;
+    
+    ViewType a("A", n, m);
+    Kokkos::deep_copy(a.d_view, 1);
+    
+    /* Covers case "Resize on Device" */
+    a.modify_device();
+    Kokkos::resize(a, factor*n, factor*m);
+    Kokkos::deep_copy(a.d_view, 1);
+    a.sync_host();
+
+    // Check device view is initialized as expected
+    scalar_type a_d_sum = 0;
+    // Execute on the execution_space associated with t_dev's memory space
+    typedef typename ViewType::t_dev::memory_space::execution_space 
+      t_dev_exec_space;
+    Kokkos::parallel_reduce(
+        Kokkos::RangePolicy<t_dev_exec_space>(0, a.d_view.extent(0)),
+        SumViewEntriesFunctor<scalar_type, typename ViewType::t_dev>(a.d_view),
+        a_d_sum);
+
+        // Check host view is synced as expected
+    scalar_type a_h_sum = 0;
+    for (size_t i = 0; i < a.h_view.extent(0); ++i)
+      for (size_t j = 0; j < a.h_view.extent(1); ++j) {
+        a_h_sum += a.h_view(i, j);
+    }
+
+    //Check
+    ASSERT_EQ(a_h_sum, a_d_sum);
+    ASSERT_EQ(a_h_sum, a.extent(0) * a.extent(1));  
+
+    /* Covers case "Realloc on Device" */
+    a.modify_host();
```

Comment:
The shrink test should verify that data is preserved.  Data is preserved if the host mod number is higher than the device...

```C++
    a.modify_host();
    Kokkos::deep_copy(a.h_view, 2);
    a.resize( n / factor, m / factor);
    a.sync_device();
```